New: Adds prefer-object-spread rule (refs: #7230)

sharmilajesupaul opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

- [X] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
This is a follow up to #7230, we implemented `prefer-object-spread` internally at @airbnb.
The rule requires that you use an object spread over an `Object.assign` call with multiple arguments and an object literal as the first argument.

We also warn on cases where an `Object.assign` call is made using a single argument that is an object literal, in this case, the `Object.assign` is not needed (7f7f97d).

The following patterns are considered errors:

```js

Object.assign({}, foo)

Object.assign({}, {foo: 'bar'})

Object.assign({ foo: 'bar'}, baz)

Object.assign({ foo: 'bar' }, Object.assign({ bar: 'foo' }))

Object.assign({}, { foo, bar, baz })

Object.assign({}, { ...baz })

// Object.assign with a single argument that is an object literal
Object.assign({});

Object.assign({ foo: bar });
```

The following patterns are not errors:

```js

Object.assign(...foo);

// Any Object.assign call without an object literal as the first argument
Object.assign(foo, { bar: baz });

Object.assign(foo, Object.assign({ bar: 'foo' }));

Object.assign(foo, { bar, baz })

Object.assign(foo, { ...baz });
```

**Is there anything you'd like reviewers to focus on?**

I made 2 commits in this PR
1.  6e3e035 handles the general use case where `Object.assign` is called using an object literal and an additional argument, which can be changed to use object spread.
2. 7f7f97d warns on cases where `Object.assign` has a single argument that is an object literal. In this case, you can use the object literal directly. (i.e. `Object.assign({})` -> `{}`)

I left the second part in its own commit because I wasn't sure if it would fit into the scope of this rule.

cc. @ljharb

mysticatea commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,180 @@
+                ) {
+                    context.report({
+                        node,
+                        message:
```
Would you use `messageId` instead `message`?

mysticatea commented on tests/lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,221 @@
+        {
+            code: "Object.assign(foo, { bar: baz })",
+            parserOptions
+        }
```
Would you add tests for `ecmaVersion: 2018`?

mysticatea commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,180 @@
+
+    create: function rule(context) {
+        return {
+            CallExpression: node => {
```
Would you use a consistent way for `create` and `CallExpression` (e.g., method shorthand)?

> I wonder if we have enabled `object-shorthand` rule in our codebase. :thinking:

mysticatea commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,180 @@
+        docs: {
+            description:
+                "disallow using Object.assign with an object literal as the first argument and prefer the use of object spread instead.",
+            category: "ECMAScript 6",
```
Would you change the category to `Stylistic Issues`?

ES6 doesn't look proper category. Though it has not done, https://github.com/eslint/eslint/issues/7991 has been accepted.

mysticatea commented on tests/lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,221 @@
+                    type: "CallExpression"
+                }
+            ]
+        }
```
Would you add tests for comments and parentheses? I expect comments to be kept.

For example:

```
Object.assign(
    {},
    // a comment
    foo,
    // another comment
    { a: true }
)
```

```
Object.assign(
    {},
    ({ a }),
    b ? c : { d },
    (e, f)
)
```

mysticatea commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,180 @@
+ * @returns {Function} autofixer - replaces the Object.assign with a spread object.
+ */
+function autofixSpread(node, sourceCode) {
+    return fixer => {
```
I guess this might be able to simplify with generator function. (It's just a impression)

sharmilajesupaul replied:
could you give me an example of this? I'm not too familiar with generator functions.

mysticatea commented on docs/rules/prefer-object-spread.md:
```diff
@@ -0,0 +1,49 @@
+
+When Object.assign is called using an object literal the first argument, this rule requires using the object spread syntax instead. This rule also warns on cases where an `Object.assign` call is made using a single argument that is an object literal, in this case, the `Object.assign` call is not needed.
+
+**Please note:** This rule can only be used when using an `ecmaVersion` of 2018 or higher, 9 or higher, or when using an `ecmaVersion` of 2015-2017 or 5-8 with the `experimentalObjectRestSpread` parser option enabled.
```
I think that the note about `experimentalObjectRestSpread` doesn't need since the option would be removed in nearly future.

platinumazure commented:
Hi @sharmilajesupaul, just wanted to follow up and ask if this is ready for another round of review or if there is more to be done here. Thanks!

sharmilajesupaul commented:
@platinumazure yes yes wow I completely forgot, sorry! I spent a bunch of time addressing multiline objects that warn on this rule and never updated my PR, I will update it today

sharmilajesupaul commented:
Made a big change to the fixer to work with multiline objects. Now it spreads object literal arguments instead of spreading them. Also addressed a majority of last code review requests. I haven't worked with generators much, so I'm not sure how that will work for the fixer? cc. @ljharb @platinumazure @mysticatea

platinumazure commented on docs/rules/prefer-object-spread.md:
```diff
@@ -0,0 +1,47 @@
+# Prefer use of an object spread over `Object.assign` (prefer-object-spread)
+
+When Object.assign is called using an object literal the first argument, this rule requires using the object spread syntax instead. This rule also warns on cases where an `Object.assign` call is made using a single argument that is an object literal, in this case, the `Object.assign` call is not needed.
```
Small typo: "using an object literal the first argument" --> "using an object literal as the first argument".

Also, we generally prefer that the first section should only have background about object spread and `Object.assign`-- the Rule Details section is where the rule's purpose/method should be outlined.

platinumazure commented on docs/rules/prefer-object-spread.md:
```diff
@@ -0,0 +1,47 @@
+
+## Rule Details
+
+The following patterns are considered errors:
```
Could this be replaced with:

```
Examples of **incorrect** code for this rule:
```

Thanks!

platinumazure commented on docs/rules/prefer-object-spread.md:
````diff
@@ -0,0 +1,47 @@
+Object.assign({ foo: bar });
+```
+
+The following patterns are not errors:
````
Could this be replaced with:

```
Examples of **correct** code for this rule:
```

Thanks!

sharmilajesupaul commented:
Just pushed updates to this PR: 
- now using `emcaVersion: 2018` in tests (https://github.com/eslint/eslint/pull/9955/files#diff-6088f3cff25518533d89b2e8a4d395faR14)

- added tests for multiple comment types and multiline objects (https://github.com/eslint/eslint/pull/9955/files#diff-6088f3cff25518533d89b2e8a4d395faR202) 

- added tests for cases where we do not need parens around an object literal (https://github.com/eslint/eslint/pull/9955/files#diff-6088f3cff25518533d89b2e8a4d395faR370) 

Sorry it took so long! 

cc @platinumazure, @mysticatea, @ljharb

platinumazure commented on docs/rules/prefer-object-spread.md:
```diff
@@ -0,0 +1,47 @@
+// Any Object.assign call without an object literal as the first argument
+Object.assign(foo, { bar: baz });
+
+Object.assign(foo, Object.assign({ bar: 'foo' }));
```
Question: Should the inner (second) `Object.assign` be flagged here? If so, would it be better to add a non-literal first argument so the whole line is a correct example?

sharmilajesupaul replied:
yeah it  should good catch

platinumazure commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,259 @@
+ */
+function addComma(arg) {
+    const nonWhitespaceCharacterRegex = /[^\s\\]/g;
+    const commentRegex = /(\/\*[\w'\s\r\n*]*\*\/)|(\/\/[\w\s']*)|(<![-\-\s\w>/]*>)/g;
```
Rather than taking a string and matching for comments, would it make more sense to use token/comment retrieval methods in `SourceCode`? I think this basically boils down to getting the last non-comment token in the argument node, and adding a comma afterward.

sharmilajesupaul replied:
Thank you for mentioning this. Instead of the regex, we can just use the ranges in `context.getCommentsInside(node)`. Definitely over complicated this a bit 😬

platinumazure commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,259 @@
+    const nonWhitespaceCharacters = Array.from(matchAll(arg, nonWhitespaceCharacterRegex));
+    const commentRanges = [];
+
+    // Create a ranges of starting and ending indicies for comments found
```
Nit: s/indicies/indices

platinumazure commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,259 @@
+    create: function rule(context) {
+        return {
+            CallExpression(node) {
+                const sourceCode = context.getSourceCode();
```
Nit: You could move this a few lines up so that we don't need to retrieve the `SourceCode` object on every CallExpression. (It's a persistent reference.)

platinumazure commented on tests/lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,422 @@
+        "let a = Object.assign(a, b)",
+        "Object.assign(a, b)",
+        "let a = Object.assign(b, { c: 1 })",
+        "let a = Object.assign({}, ...b)",
```
Should this be valid? Wondering if the correct code should be `let a = Object.assign(...b);`, or am I missing something? (Compare with the test case 2 lines below this one)

ljharb replied:
Statement and expression positions differ in terms of how the autofixer outputs parens, but good call, this one should be a spread.

platinumazure commented on tests/lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,422 @@
+            ]
+        },
+
+        // TODO: Handle nested Object.assign calls
```
Should this comment be removed? (Although, now I see some of the test cases below, seems there might be more to do here?)

platinumazure commented on tests/lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,422 @@
+            errors: [
+                {
+                    messageId: "useSpreadMessage",
+                    type: "CallExpression"
```
Could this error (and the one below) have more information to help identify where the errors are reported? (e.g., line/column)

platinumazure commented on tests/lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,422 @@
+        {
+            code:
+                "Object.assign({ foo: 'bar' }, Object.assign({ bar: 'foo' }, Object.assign({}, { superNested: 'butwhy' })))",
+            output: "({foo: 'bar', ...Object.assign({ bar: 'foo' }, Object.assign({}, { superNested: 'butwhy' }))})",
```
I think the innermost `Object.assign` should be identified and fixed-- is that a known limitation at this point? Or are we relying on multipass autofix to catch these cases? (Now the TODO above makes more sense...)

ljharb replied:
Multipass, i think.

sharmilajesupaul replied:
I had a solution earlier that would recurse through nested object.assign calls, but the current implementation relies on multipass

platinumazure commented on tests/lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,422 @@
+            errors: [
+                {
+                    messageId: "useSpreadMessage",
+                    type: "CallExpression"
```
Could this error (and the ones below) have more information, e.g. line/column numbers?

sharmilajesupaul commented:
@platinumazure ptal, made improvments 
https://github.com/eslint/eslint/pull/9955/commits/7a7960ad38102f51a0ecf67a25d1ee29f9143c9f

platinumazure commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,249 @@
+        schema: [],
+        fixable: "code",
+        messages: {
+            useSpreadMessage: "Use an object spread instead of `Object.assign()` eg: `{ ...foo }`",
```
I don't think the example should be necessary. Users can view the documentation (or search the Internet) for what object spread means.

platinumazure replied:
Well, turning that around, why doesn't the object literal case have an example?

Also: Since we have autofix here, the user could just let that be run and see how the item should look. (And no, we don't need to note that `--fix` could be run in the lint message, because we note that in most formatters outside of the lint message)

ljharb replied:
I suppose that's a fair point - maybe it should have an example :-)

(also since autofix can't be easily run for a single rule, that's not really a practical option for demonstrating the results of a single rule)

sharmilajesupaul commented:
@platinumazure there should have been a line in the object literal case, to make sure that it was only warning on `Object.assign`, not sure how I missed that. But I added it back in along with the test case: 
https://github.com/eslint/eslint/pull/9955/commits/751774deccdfd030b80e214ad6f17c66db0a3ac3#diff-c7ab2d044a2d5cc255c07ea6116488e1R267

I also updated the rule to pick up on whether the native `Object` is overwritten in a variable declaration or assignment in the current file. https://github.com/eslint/eslint/pull/9955/commits/751774deccdfd030b80e214ad6f17c66db0a3ac3#diff-c7ab2d044a2d5cc255c07ea6116488e1R210

not-an-aardvark commented on lib/rules/prefer-object-spread.js:
```diff
@@ -0,0 +1,278 @@
+ */
+function addComma(formattedArg, comments) {
+    const nonWhitespaceCharacterRegex = /[^\s\\]/g;
+    const nonWhitespaceCharacters = Array.from(matchAll(formattedArg, nonWhitespaceCharacterRegex));
```
Is it necessary to use `matchAll` here? Since this regex doesn't have any capturing groups, it seems like this could just be done with:

```js
const nonWhitespaceCharacters = formattedArg.match(nonWhitespaceCharacterRegex);
```

That would avoid the need to add a dependency on `string.prototype.matchall`.

ljharb replied:
With a global regex, `.match` does not return an array of match objects, so there's no way to get the index (position) of each match - this is the entire reason `.matchAll` is needed.

So, no, I don't think there's any way to avoid the dependency (without inlining it, of course, which would be a terrible idea)
